solr/core/src/java/org/apache/solr/core/SolrCores.java

```diff
@@ -536,7 +538,9 @@ public void queueCoreToClose(SolrCore coreToClose) {
    * @return the cache holding the transient cores; never null.
    */
   public TransientSolrCoreCache getTransientCacheHandler() {
```

Comment:
I think it should be a prerequisite that the caller acquire the lock a-priori.  For example, close() seems to need to keep holding this lock to call close() on it.

Reply:
I don't think so. Let's say transientSolrCoreCacheFactory = A. If there is a race and load() is called between getTransientCacheHandler() and TransientSolrCoreCache.close(). getTransientCacheHandler() returns A, then load() sets B, and then A.close() is called. It is the same result as the sequence { getTransientCacheHandler().close() } atomically on A, and then load() B.

But actually it doesn't matter. So I can add a synchronized (modifyLock) around if you prefer.
I'll keep anyway the synchronized inside getTransientCacheHandler() because it is public.

Reply:
Okay; I'm fine with the lock to be safe.